File: 61-rotate-list/rotate-list.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def rotateRight(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        if not head : return head
        length = 1
        tail = head
        while tail.next:
            tail = tail.next
            length+=1
          # Compute the effective number of rotations needed, 
        # as rotating a list by its length results in the same list
        k = k % length
        
        # If no rotation is needed, return the head as it is
        if k == 0:
            return head
        curr = head
        for i in range(length -k -1):
            curr = curr.next
        tmp = curr.next
        curr.next = None
        tail.next = head

        return tmp



        

File: 61-rotate-list/rotate-list.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} k
 * @return {ListNode}
 */
var rotateRight = function(head, k) {
    if (!head || !head.next || k === 0) return head;

    // Step 1: Get the length and make it circular
    let length = 1;
    let tail = head;
    while (tail.next) {
        tail = tail.next;
        length++;
    }
    tail.next = head; // Form a circle

    // Step 2: Find the new tail position
    k = k % length;
    let stepsToNewTail = length - k;
    let newTail = head;
    for (let i = 1; i < stepsToNewTail; i++) {
        newTail = newTail.next;
    }

    // Step 3: Break the circle and return new head
    let newHead = newTail.next;
    newTail.next = null;
    return newHead;
};